### PR TITLE
Hash upgrade

### PR DESCRIPTION
**Pull request by BukrosSzabolcs**
Switch to crypt(3) hashing when storing user passwords in zk
Extending authentication to recognize users with outdated hash and after a warning at startup update them on successful authentication.

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:65]: since we aren't including any details about which user(s) are impacted, I'd rather we not get a WARN for each user. could we move this to after we finish iterating with a summary of how many users?

If an operator needed to move towards eliminating these warn messages, how would they get the list of users that need to authenticate to the system?

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:70]: log a DEBUG message with these details.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:140]: could we log a WARN message that this method has been used? that way it would show up in operator logs should we mistakenly use it in a non-test context.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:259]: if we fail to update the password for some reason (like a transient zk write failure), at this point shouldn't we still return that they correctly authenticated?

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:52]: add a comment about why a new Charset instead of relying on `StandardCharsets.UTF_8`

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:68]: just to make sure I understand the reasoning here, the removal of this check is because ZKSecurityTool is package private and all current calls ensure zkData isn't null?

if that's correct please add javadocs that say zkData can't be null. an alternative is to leave the check in place and rely on the jit to optimize it away.

File: core/src/main/java/org/apache/accumulo/core/Constants.java
Comment (busbey) [inline core/src/main/java/org/apache/accumulo/core/Constants.java:102]: We still need this constant updated because we rely on it for hashing the system credentials? won't that prevent a rolling upgrade?

Could we have system credentials fall back to SHA-256 with a warning? or require a configurable flag to switch it?

Comment (busbey) [reply core/src/main/java/org/apache/accumulo/core/Constants.java:102]: changing this is also going to change some non-security uses, e.g. we optionally use it to obscure values printed from rfile metrics gathering. We'll need to enumerate these and release note the change in behavior. (or we could make something like a `NON_CRYPTO_USE_HASH_ALGORITHM` that we keep as SHA-256)

**Review by busbey — CHANGES_REQUESTED** [review]

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:120]: What's the advantage of using `MessageDigest.isEqual`, vs. comparing as Strings or comparing as the UTF-8 byte-array?

File: server/base/src/test/java/org/apache/accumulo/server/security/handler/ZKAuthenticatorTest.java
Comment (ctubbsii) [inline server/base/src/test/java/org/apache/accumulo/server/security/handler/ZKAuthenticatorTest.java:101]: You can simplify this JUnit test by just throwing this out of the method. JUnit will show the details of the exception in its logs.

File: test/src/main/java/org/apache/accumulo/test/ZKAuthenticatorIT.java
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/ZKAuthenticatorIT.java:30]: If we can avoid running a full ZK instance using Mini, I think we should. I think ZKAuthenticator can be tested fully using unit tests with mocking, without a need to start up an instance of a cluster.

**Review by ctubbsii — COMMENTED** [review]
This is great, @BukrosSzabolcs ! I think this is like 95% done now. Thanks! There's just a few comments to address, and then I'll probably want to run some manual upgrade tests just as a sanity check for myself, but it looks like we'll have no problem getting this merged in for 2.1.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:65]: I'm never sure exactly how much detail is too much when it comes to logging security features. I can add an affected user count.
I'm not sure the operator has to or should do anything, the update is automatic. But with their current tools they could replicate what the code does and could go through the users zNode checking hash lengths. Alternately I could list the affected user principals in the warning but that feels unsafe. Do you have something in mind?

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:52]: Switching to `StandardCharsets.UTF_8`

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:120]: String comparisons is not secure against timing attacks, `MessageDigest.isEqual` is.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:65]: The list of users in the server logs is okay. However, I just realized... this method is only ever called when we're starting from an empty user database (on initialize... or re-initialize), so there shouldn't be any matching the old hashes anyway at the point this code is run. Or did I miss something?

I do think maybe it would be worth the master server checking to see if any users are using the old hashing, and listing them in the server logs on startup, as an upgrade warning. In future, we can turn that into an upgrade blocker (so we can eventually drop the code that supports reading the old hashes).

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:68]: I re-added the check to be on the safe side, but and made it package private as I should have done already.

File: server/base/src/test/java/org/apache/accumulo/server/security/handler/ZKAuthenticatorTest.java
Comment (BukrosSzabolcs) [reply server/base/src/test/java/org/apache/accumulo/server/security/handler/ZKAuthenticatorTest.java:101]: Thank you for the suggestion!

File: core/src/main/java/org/apache/accumulo/core/Constants.java
Comment (BukrosSzabolcs) [reply core/src/main/java/org/apache/accumulo/core/Constants.java:102]: You are right. I'll make the system credentials hash customizable while defaulting to 256,  and renaming this as suggested to reflect it's for unsecure use.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:70]: It looks like initialize itself is already an edge case... see my comment at https://github.com/apache/accumulo/pull/1798/files/4c8fedce90487c40ec4b740bc98dc181177cc71f#r528903518

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:65]: Added the check to master startup through SecurityOperation.

File: test/src/main/java/org/apache/accumulo/test/ZKAuthenticatorIT.java
Comment (BukrosSzabolcs) [reply test/src/main/java/org/apache/accumulo/test/ZKAuthenticatorIT.java:30]: It can. I'm rewriting it.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java]: @BukrosSzabolcs Feel free to mark the conversation as "Resolved" if you've taken steps to address the comments. It cleans up the interface and makes it easier for subsequent reviews. :smiley_cat:

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:89]: See my other comment about not making this configurable. However, since it is configurable, it must be the same across all servers, or they won't be able to talk to each other. In general, that would mean making this an `instance.*` property. However, since the way we enforce these is through the same-ness of the system credentials, if this is different across the cluster, then they won't be able to talk to each other anyway. The other property prefix we use for system-wide properties that can be different are `general.*`.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:139]: Since this use of the digest is specific to this specific metric utility's serialization, we can probably just use a local constant, rather than one in `Constants.java` (which I'd personally like to phase out over time).

File: shell/src/main/java/org/apache/accumulo/shell/commands/GetSplitsCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/GetSplitsCommand.java:116]: It looks like we have this exact same algorithm for creating obscured tablet identifiers implemented in at least two places. Creating a new issue (#1805) to address that separately.

File: server/base/src/main/java/org/apache/accumulo/server/security/SystemCredentials.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/SystemCredentials.java:106]: I would prefer to leave this as it is for now and create a separate issue for the follow up changes and related discussion.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:62]: I considered that too, but kept it like this because we are checking for the outdated format, and while not being outdated technically means it's the current one, the name still feels misleading.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java
Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java:54]: Sure, thanks for pointing it out.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:152]: yeah that sounds great.

Comment (BukrosSzabolcs) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:152]: Just realized the only reason for this method was that the IT tests required it, but with reworking those to unit tests this is no longer used. So I'll just remove this.

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:271]: I think this comes down to wether we want to fail open or closed for updating to a more secure hash. I personally prefer failing open in this case -- essentially if there's a problem writing to ZK we did not consider here then the system will keep working as before the upgrade but with a persistent log message an operator could use to go fix things.

I could see at least one argument for failing closed: a person or system not being able to talk to Accumulo will be a very loud noise that will ensure whatever is preventing us from storing things more securely gets handled. Presuming secure hashing of password data is an important system requirement in a particular Accumulo deployment that is a good thing.

Does anyone else feel strongly?

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (busbey) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:89]: I personally think `instance` seems an appropriate prefix, so it would be `instance.system.token.hash.type`.

Comment (BukrosSzabolcs) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:89]: Thank you for the suggestions. Adding the `instance.*` prefix.

**Comment by BukrosSzabolcs** [comment]
The test failure looks unrelated. How can I re-trigger the tests?

**Comment by Manno15** [comment]
They all seemed to pass this time.

**Comment by BukrosSzabolcs** [comment]
@Manno15 Thanks for your help!

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:246]: I don't feel very strongly about it. However, I do lean towards forcing the upgrade and rejecting if it can't be completed (for the reasons @busbey mentioned, as well as because I want to eventually ensure we can drop the code that supports reading the old hashes on some future upgrade).
I think it's very unlikely we're going to fail here if we've already authenticated, unless ZK is having issues (in which case, we'll probably see other errors in the system). So, it may be moot. Either way would probably be fine.

File: server/base/src/main/java/org/apache/accumulo/server/security/SystemCredentials.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/security/SystemCredentials.java]: No problem. I created #1810 to deal with it as a follow-on issue.

**Review by ctubbsii — COMMENTED** [review]
These changes look good to me, but I want to do a bit of manual testing of them before merging, if that's okay. (also, @busbey has an outstanding request, so we'll wait on his approval also, before merging)

File: server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java
Comment (busbey) [reply server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java:246]: okay, sounds good. @BukrosSzabolcs I think you're fine to leave this as-is.

**Review by busbey — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:49]: This is an odd one.  I am not sure what we are using the HASH to encrypt here.  If it is just to prevent leaking data then I _think_ it would be OK to use the NON_CRYPTO_USE_HASH_ALGORITHM

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:49]: > This is an odd one. I am not sure what we are using the HASH to encrypt here. If it is just to prevent leaking data then I _think_ it would be OK to use the NON_CRYPTO_USE_HASH_ALGORITHM

@milleruntime I think this was moved here at my request to keep serialization-related constants local to the serialization code, rather than to rely on some global constant, which could have a multitude of uses. I think it's fine here.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:49]: I don't think this is serializing anything.  I thought this method was only used by the `rfile-info` command. The only use i can find of `printMetircs()` is here: https://github.com/apache/accumulo/blob/6a74b4667e3bd33e34b5262c5dd8ea64167fb657/core/src/main/java/org/apache/accumulo/core/file/rfile/PrintInfo.java#L291

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:49]: By serialization, I simply mean the serial form that we use for this particular display format.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:49]: That constant can be left inside this class, because it's not being used as some sort of overall crypto strategy that necessitates it be aligned with how the other constant is being used... it's just an arbitrary choice for this particular visibility metric output format (or "serialization algorithm") and relates to nothing else. So, it can/should be a local constant held closer to the rest of that metric serialization algorithm. That's all I mean.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/VisMetricsGatherer.java:49]: Ah I see.  It makes me wonder why are we even bother hashing it at all for display purposes.  Why not just print gobbledygook or just don't print it.

File: server/manager/src/main/java/org/apache/accumulo/master/Master.java
Comment (milleruntime) [inline server/manager/src/main/java/org/apache/accumulo/master/Master.java:1182]: I am wondering if this is the best place to validate all user creds.  On one hand, it may be better to do this for some situations every time Master starts up but on the other, it may be sufficient and more efficient to only check once on upgrade.

Comment (ctubbsii) [reply server/manager/src/main/java/org/apache/accumulo/master/Master.java:1182]: I was wondering that, too, but I kind of want users to be annoyed by it more than once, so that they actually log users in to upgrade the hashes. I was also thinking maybe the other route, and ~~nagging~~ *logging* once every few days or once a week on a running system. Once at master startup is probably a good compromise for now, though. We can get more aggressive in future releases, before eventually forcing it at upgrade.

Comment (milleruntime) [reply server/manager/src/main/java/org/apache/accumulo/master/Master.java:1182]: Good point.  And we can easily move the check somewhere else in later versions.

**Review by milleruntime — APPROVED** [review]
Thanks for the work on this change @BukrosSzabolcs !!

**Comment by BukrosSzabolcs** [comment]
@ctubbsii The outstanding request was resolved and got the approval. Have you had time to do some manual testing?

**Comment by ctubbsii** [comment]
> @ctubbsii The outstanding request was resolved and got the approval. Have you had time to do some manual testing?

Not yet. This is my priority today.

**Comment by ctubbsii** [comment]
@BukrosSzabolcs I added a commit to merge the main branch onto this, to get it up-to-date with current main, and to resolve compiler warnings, but I'm still testing. I should be done today.

**Review by ctubbsii — COMMENTED** [review]
This looks great to me. I did some manual testing, and everything works as expected. I do think it would be useful to see an INFO (maybe DEBUG? any opinions here?) message when each user's credentials are upgraded, though.

Also, I noticed some strange "denied" messages for "performSystemAction" when a user successfully authenticates. I don't know if that's new with this PR or not. I'm going to see if I can track it down. If it's pre-existing, then I wouldn't worry about it.

**Comment by ctubbsii** [comment]
> This looks great to me. I did some manual testing, and everything works as expected. I do think it would be useful to see an INFO (maybe DEBUG? any opinions here?) message when each user's credentials are upgraded, though.

Don't worry about this one. I will add a commit to add a log message for the upgrade action before merging. Everything here looks great! I will merge it soon. Thanks, @BukrosSzabolcs !

> Also, I noticed some strange "denied" messages for "performSystemAction" when a user successfully authenticates. I don't know if that's new with this PR or not. I'm going to see if I can track it down. If it's pre-existing, then I wouldn't worry about it.

It looks like this is a pre-existing issue. Don't worry about it.

**Comment by BukrosSzabolcs** [comment]
Thanks for the feedback and the help @ctubbsii, @busbey !

**Comment by ctubbsii** [comment]
@BukrosSzabolcs I ran into a problem while testing, because I made a mistake on #1827 , but I reverted that, and was able to finally merge this change in (with my added debug log statement, as discussed).

**Comment by ctubbsii** [comment]
Thank you so much for your contribution! I'm really happy we were finally able to make this change!

**Pull request merged**
